### PR TITLE
Debugger widget

### PR DESCRIPTION
**Pull request by hyamanieu**
Added debugger widget. Fixes https://github.com/holoviz/panel/issues/2062
NB: only interesting change in reactive.py is within `_process_events`
Sample code to try out:

```python
import panel as pn
import param
pn.extension()
class A(param.Parameterized):

    with_throttled_enabled = param.Range(
        default=(100, 250),
        bounds=(0, 250),
    )

    def __init__(self, **params):
        super().__init__(**params)
        widgets = {
            "with_throttled_enabled": {
                "type": pn.widgets.IntRangeSlider,
                "throttled": False,
            },
        }
        self.controls = pn.Param(self, widgets=widgets)

        
    @param.depends("controls")
    #@status_decorator(behavior='erase')
    def calculation(self):
        if self.with_throttled_enabled[0] > 130:
            raise ValueError('too low')
        return pn.Pane((self.with_throttled_enabled), min_width=200)

    @param.depends("controls")
    #@status_decorator(behavior='erase')
    def buggy_calculation(self):
        return pn.Pane(self.with_throttled_enabled[1]/self.with_throttled_enabled[0], min_width=200)

a = A()


debug = pn.widgets.Debugger(name='Debugger')


a = pn.Column(a.controls, pn.pane.Markdown('### will bug above 130'), 
          a.calculation, 
          pn.pane.Markdown('### division by 0'), 
          a.buggy_calculation, 
          debug,
          )
a.servable()
```

## To-do:
- [x] Correct visualization defaults
- [x] Correct button locations
- [x] Add example notebooks for the doc
- [x] Add test
- [x] make sure the console output works as usual in a notebook.

### CODE REVIEW

File: panel/reactive.py [hunk 1]
```diff
@@ -253,22 +249,20 @@ def _param_change(self, *events):
...
-                self.param.set_param(**self_events)
-            for k, v in self_events.items():
-                if '.' not in k:
-                    continue
-                *subpath, p = k.split('.')
-                obj = self
-                for sp in subpath:
-                    obj = getattr(obj, sp)
-                with edit_readonly(obj):
-                    obj.param.set_param(**{p: v})
```
Comment (philippjfr) [inline panel/reactive.py:271]: You undid a bunch of changes here.

File: panel/reactive.py [hunk 2]
```diff
@@ -290,19 +284,15 @@ def _change_event(self, doc=None):
             state.curdoc = None
             state._thread_id = None
 
-    def _comm_change(self, doc, ref, comm, subpath, attr, old, new):
-        if subpath:
-            attr = f'{subpath}.{attr}'
```
Comment (philippjfr) [inline panel/reactive.py:295]: And here.

File: panel/reactive.py [hunk 3]
```diff
@@ -290,19 +284,15 @@ def _change_event(self, doc=None):
...
         if attr in self._changing.get(ref, []):
             self._changing[ref].remove(attr)
             return
 
         with hold(doc, comm=comm):
             self._process_events({attr: new})
 
-    def _server_change(self, doc, ref, subpath, attr, old, new):
-        if subpath:
-            attr = f'{subpath}.{attr}'
```
Comment (philippjfr) [inline panel/reactive.py:305]: And here.

**Comment by philippjfr** [comment]
This looks great! However it looks like you started with an old(er) checkout and simply undid a bunch of recent changes when you rebased.

**Comment by hyamanieu** [comment]
Yes I have noticed the mistake but didn't have time to fix. I added a
comment about it, I'll fix the merge conflict next time.

Le mar. 20 juil. 2021 à 16:10, Philipp Rudiger ***@***.***> a
écrit :

**Comment by hyamanieu** [comment]
I can't get to have an exception accumulated in a notebook cell, even with master branch & 0.11.3. Not sure how `ServableMixing._on_error` gets triggered.

**Comment by philippjfr** [comment]
Would be great to take this opportunity to add a bunch more log statements throughout the codebase as. The places you identified are definitely good but there's probably a bunch more that would help.

**Comment by hyamanieu** [comment]
> Would be great to take this opportunity to add a bunch more log statements throughout the codebase as. The places you identified are definitely good but there's probably a bunch more that would help.

I identified the functions called whenever a callback is fired. 

I would use `panel.callbacks` logger only when the end user is the target, not the Panel developers. Or what is your intent?

File: panel/reactive.py (hunk 3 above)
Comment (hyamanieu) [reply panel/reactive.py]: outdated

**Comment by philippjfr** [comment]
>I would use panel.callbacks logger only when the end user is the target, not the Panel developers. Or what is your intent?

No that makes sense. Just wondering if we can have a few log levels. I'm also thinking of adding an `/admin` endpoint to a server if requested, which provides details about the active sessions and optionally provides access to detailed logs as well.

**Comment by hyamanieu** [comment]
> > I would use panel.callbacks logger only when the end user is the target, not the Panel developers. Or what is your intent?
> 
> No that makes sense. Just wondering if we can have a few log levels. I'm also thinking of adding an `/admin` endpoint to a server if requested, which provides details about the active sessions and optionally provides access to detailed logs as well.

A bit like in jupyterhub, that could be helpful indeed. But that could be left at the discretion of the developing user, all the tools are already there to do such a thing.

**Comment by philippjfr** [comment]
Sorry I left this so late to review, I've cleaned this up a bit and changed it to subscribe to the main panel logger. I'm still a little bit concerned about the brittleness of the width/height/sizing_mode handling which was pretty broken when I first looked at this. I've had to make some changes to terminal sizing but I don't think this handles every combination of min/max_width/height options and different sizing modes correctly. I do want to get a 0.13.0 RC release out today. Will try to clean this up more and get it in.

**Comment by hyamanieu** [comment]
Philipp, thanks for your review & corrections.
I have a bettered version of this widget in a separate project, I still didn't have the time to add it within this branche of Panel and won't have time for the next two weeks. 

Some features that my experience with my user case made me add:

- I have fixed some sizing problems between card & terminal, although I may have not thoroughly tested it like you as I simply fixed it.
- I have added the possibility to optionally subscribe to additional loggers
- I have added a third button to send an email with logs attached to a specified address (basically a `mailto:` URI)
- I have added an option to have a flying debugger if embedded within a template (it requires changes in the template css however :s )

To do:
- [ ] Make it work in jupyter. After pulling from master yesterday, it's not working in jupyter (trad and lab) anymore. I have unticked it from my first message.
- [ ] (optionally) find a way to auto-test widgets within notebook if it's possible ; so that next time we don't need to manual test it.


There are still some problems with terminal itself where opening and closing the card. The text sometimes appear and sometimes disappear. Interestingly, exporting the log by clicking on the floppy disk button shows the full output even if the terminal is completely black. After testing a bit more, I will open a separate issue as it's not related to this widget but the interaction of `Terminal `& `Card`.

Besides the js console keeps complaining about the missing buttons when the card is closed. Not sure if a callback can be "deactivated" whenever the Card is closed so that we don't have the warning in the console.

**Comment by hyamanieu** [comment]
@philippjfr I have tested your changes. Very nice. I could not make sizing errors and the debugger now works in a jupyter context.

I have added a couple options (logging formatting + decide which logger to prompt) and fixed documentation.

I allowed myself to ping you for a review. There is a left over `print`, not sure this was wanted.

Besides the latter, to me it is ready to merge.

Thanks :)
ps: I'll add the "flying" effect and the mailing option at some point later. Current version works already well for end users to communicate clearly with dashboard developers.

**Pull request merged**